Buttplug.Core/ButtplugJsonMessageParser.cs

```diff
@@ -152,22 +157,54 @@ public ButtplugMessage[] Deserialize(string aJsonMsg)
             return res.ToArray();
         }
 
-        public string Serialize([NotNull] ButtplugMessage aMsg)
+        public string Serialize([NotNull] ButtplugMessage aMsg, uint clientSchemaVersion)
         {
             // Warning: Any log messages in this function must be localOnly. They will possibly recurse.
-            var o = new JObject(new JProperty(aMsg.GetType().Name, JObject.FromObject(aMsg)));
+
+            // Support downgrading messages
+            var tmp = aMsg;
+            while (tmp.MessageVersioningVersion > clientSchemaVersion)
+            {
+                if (tmp.MessageVersioningPrevious == null)
+                {
+                    // There's no previous version of this message
+                    // Not sure what to do here... send an empty array?

```

Comment:
If we're trying to send a message that the client won't understand, an empty array won't do us much good. Maybe create a new Protocol error class and send an error with that?

Reply:
This shouldn't really happen. I guess if there's a non-zero message ID we should send back something, otherwise drop the send entirely (new server initiated messages that the client doesn't understand are not going to be missed).